modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,733 @@
+proc isDiag(A: [?D] ?eltType) {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+
+  const zero = 0: eltType;
+  for (i, j) in D {
+    if i != j && A[i, j] != zero then return false;
+  }
+  return true;
+}
+
+
+/* Return `true` if matrix is Hermitian */
+proc isHermitian(A: [?D]) {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  for (i, j) in D {
+    if i > j {
+      if A[i, j] != conjg(A[j, i]) then return false;
+    }
+  }
+  return true;
+}
+
+
+/* Return `true` if matrix is symmetric */
+proc isSymmetric(A: [?D]) : bool {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  for (i, j) in D {
+    if i > j {
+      if D[i, j] != D[j, i] then return false;
+    }
+  }
+  return true;
+}
+
+
+/* Return `true` if matrix is lower triangular below the diagonal + ``k`` */
+proc isTril(A: [?D] ?eltType, k=0) : bool {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  const zero = 0: eltType;
+  for (i, j) in D do
+    if (i < j-k) && (A[i, j] != zero) then
+      return false;
+  return true;
+}
+
+
+/* Return `true` if matrix is upper triangular */
+proc isTriu(A: [?D] ?eltType, k=0) : bool {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  const zero = 0: eltType;
+  for (i, j) in D do
+    if (i > j-k) && (A[i, j] != zero) then
+      return false;
+  return true;
+}
+
+
+/* Return `true` if matrix is square */
+proc isSquare(A: [?D]) {
+  if D.rank != 2 then
+    compilerError("Rank size is not 2");
+  const (M, N) = A.shape;
+  return M == N;
+}
+
+
+/* Return the determinant of square matrix ``A`` */
+proc det(A: [?D] ?eltType) {
+  // Naive algorithm -
+  //    Better solution uses LU decomp: det M = det LU = det L * det U
+  if D.rank != 2 then
+    compilerError("Rank size not 2");
+
+  // Handle odd domains inefficiently:
+  if D.stridable {
+    var Atmp = Matrix(A);
+    return det(Atmp);
+  }
+  if D.first != 0 || D.last != D.shape(1) {
+    var Atmp = Matrix(A);
+    return det(Atmp);
+  }
+  // Assume 0-based non-strided domain below this point
+
+  if !isSquare(A) then
+    halt('non-square array passed to determinant. Aborting');
+
+  const n = D.shape(1);
+  if n == 1 then return A[0];
+
+  var d = 0: eltType;
+  forall j in 0..#n with (+ reduce d) do
+    d += (((-1)**(j))*(A[0,j])*det(_minor(A, j))): eltType;
+
+  return d;
+}
+
+
+pragma "no doc"
+/* Helper function for determinant */
+private proc _minor(A: [?Adom] ?eltType, j) {
+  const n = Adom.dim(1).size;
+  var a : [0..#n-1, 0..#n-1] eltType;
+  var K = 0,
+      L = 0,
+      i = 0;
+
+  for k in 0..#n {
+    for l in 0..#n  {
+      // K
+      if k == i || l == j then continue;
+      if k < i then K = k;
+      else K = k-1;
+      // L
+      if l < j then L = l;
+      else L = l-1;
+      // a[K, L]
+      a[K,L] = A[k,l];
+    }
+  }
+  return a;
+}
+
+
+/* Return the trace of ``A`` */
+proc trace(A: [?D] ?eltType) {
+  if D.rank == 1 then return + reduce A[..];
```

Comment:
Hm.... the rank-1 behavior isn't consistent with the rank-2 behavior. 

- I think the rank-2 case only works for a square matrix (I'm fine with that), but it's probably worth an explicit check, as opposed to letting ``zip`` throw an error.
- In that case, the rank-1 case should only be allowed for a single element vector....

I'd be fine dropping the rank 1 case altogether.